src/Perl6/Metamodel/RoleToClassApplier.nqp

```diff
@@ -7,7 +7,18 @@ my class RoleToClassApplier {
             return nqp::existskey(%mt, $name);
         }
         else {
-            for $target.HOW.mro($target) {
+
+            # In an uncomposed class we need to look for methods in
+            # the part of the mro that has not been added yet.
+            my @mro := $target.HOW.mro($target);
+            if $target.HOW.parents($target, :local(1)) == 0
+              && $target.HOW.has_default_parent_type
+              && $target.HOW.name($target) ne 'Mu' {
+                my $dt := $target.HOW.get_default_parent_type;
+                nqp::splice(@mro, $dt.HOW.mro($dt), nqp::elems(@mro), 0);
```

Comment:
This modifies the `@mro` array in-place, but the `mro` method that computes it also, I believe, caches the result. Since `has_method` is called many times, then I suspect this ends up adding the default parent type each time so the thing in `mro` is not only "polluted" but also gets longer. I'd suggest:

1. Do an `nqp::clone` on the result of calling `$target.HOW.mro($target)`
2. Move that logic out of `has_method` and into the caller of it, and pass in an `@mro` parameter, so we only have to do the work once

These details aside, the patch seems reasonable.

Reply:
As far as I can tell, though the mro method does cache that, the cached copy is either A) on a call to a role that has not been composed in which case it is blown away (ClassHOW line 104) by a new call to compute_mro immediately thereafter, or B) never gets altered because a composed class is either Mu or has parents.  But yeah the loop over the @collisions and %methods would accumulate extra cruft.

I guess the EnumHOW also uses this.

Anyway, is there a reason not to just move the .add_parent call in ClassHOW above the whole role manipulation, and just hold off on setting $!composed until after roles are done?  That would be even more efficient... the only reason Any/Mu are missed by the role applier is that they are not in the .mro.